server/src/main/java/org/apache/cassandra/sidecar/coordination/SidecarPeerHealthMonitorTask.java

```diff
@@ -0,0 +1,191 @@
+
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import io.vertx.core.CompositeFuture;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import org.apache.cassandra.sidecar.cluster.instance.InstanceMetadata;
+import org.apache.cassandra.sidecar.codecs.SidecarInstanceCodecs;
+import org.apache.cassandra.sidecar.common.client.SidecarInstance;
+import org.apache.cassandra.sidecar.common.server.utils.DurationSpec;
+import org.apache.cassandra.sidecar.config.SidecarConfiguration;
+import org.apache.cassandra.sidecar.config.SidecarPeerHealthConfiguration;
+import org.apache.cassandra.sidecar.tasks.PeriodicTask;
+import org.apache.cassandra.sidecar.tasks.ScheduleDecision;
+import org.jetbrains.annotations.NotNull;
+
+import static org.apache.cassandra.sidecar.server.SidecarServerEvents.ON_SIDECAR_PEER_DOWN;
+import static org.apache.cassandra.sidecar.server.SidecarServerEvents.ON_SIDECAR_PEER_UP;
+
+/**
+ * Pings other 'peer' Sidecar(s) that are relevant to this Sidecar over HTTP and notifies
+ * listeners when other Sidecar(s) goes DOWN or OK.
+ */
+@Singleton
+public class SidecarPeerHealthMonitorTask implements PeriodicTask
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(SidecarPeerHealthMonitorTask.class);
+
+    private final Vertx vertx;
+    private final SidecarPeerHealthConfiguration config;
+    private final SidecarPeerProvider sidecarPeerProvider;
+    private final SidecarPeerHealthProvider healthProvider;
+
+    private final Map<SidecarInstance, SidecarPeerHealthProvider.Health> status = new ConcurrentHashMap<>();
+
+    @Inject
+    public SidecarPeerHealthMonitorTask(Vertx vertx,
+                                        SidecarConfiguration sidecarConfiguration,
+                                        SidecarPeerProvider sidecarPeerProvider,
+                                        SidecarPeerHealthProvider healthProvider,
+                                        SidecarInstanceCodecs sidecarInstanceCodecs)
+    {
+        this.vertx = vertx;
+        this.config = sidecarConfiguration.sidecarPeerHealthConfiguration();
+        this.sidecarPeerProvider = sidecarPeerProvider;
+        this.healthProvider = healthProvider;
+        vertx.eventBus().registerDefaultCodec(SidecarInstance.class, sidecarInstanceCodecs);
+    }
+
+    @NotNull
+    public SidecarPeerHealthProvider.Health status(InstanceMetadata instance)
+    {
+        return status.getOrDefault(instance, SidecarPeerHealthProvider.Health.UNKNOWN);
+    }
+
+    @Override
+    public ScheduleDecision scheduleDecision()
+    {
+        return config.enabled() ? ScheduleDecision.EXECUTE : ScheduleDecision.SKIP;
+    }
+
+    @Override
+    public DurationSpec delay()
+    {
+        return config.executeInterval();
+    }
+
+    @Override
+    public void execute(Promise<Void> promise)
+    {
+        try
+        {
+            run().onComplete(v -> promise.tryComplete()).wait(10000);
```

Comment:
1. `.wait(10000)` is not wanted. Can you remove it? You do not want to block the thread. 
2. `onComplete(v -> promise.tryComplete())` is missing the failure from `run()`.


```suggestion
            run()
            .onSuccess(v -> promise.tryComplete())
            .onFailure(promise::tryFail);
```